File: backend/app/api/websocket.py
"""
WebSocket Routes
================
Real-time bidirectional communication for live preview streaming
and shard activation visualization.
"""
import asyncio
import json
import logging
from dataclasses import asdict
from typing import Any
from uuid import uuid4

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from ..core.engine import (
    HolographicReconstructionEngine,
    ReconstructionMode,
    ReconstructionProgress,
)
from ..core.self_healing import HealthStatus, SelfHealingWatchdog

logger = logging.getLogger(__name__)
router = APIRouter()


class ConnectionManager:
    """Manages WebSocket connections for broadcasting updates."""

    def __init__(self):
        self.active_connections: dict[str, WebSocket] = {}
        self.job_subscriptions: dict[str, set[str]] = {}  # job_id -> connection_ids

    async def connect(self, websocket: WebSocket) -> str:
        """Accept connection and return connection ID."""
        await websocket.accept()
        connection_id = str(uuid4())
        self.active_connections[connection_id] = websocket
        logger.info(f"WebSocket connected: {connection_id}")
        return connection_id

    def disconnect(self, connection_id: str) -> None:
        """Remove connection."""
        if connection_id in self.active_connections:
            del self.active_connections[connection_id]
            # Remove from all job subscriptions
            for job_subs in self.job_subscriptions.values():
                job_subs.discard(connection_id)
            logger.info(f"WebSocket disconnected: {connection_id}")

    def subscribe_to_job(self, connection_id: str, job_id: str) -> None:
        """Subscribe connection to job updates."""
        if job_id not in self.job_subscriptions:
            self.job_subscriptions[job_id] = set()
        self.job_subscriptions[job_id].add(connection_id)

    async def send_personal(self, connection_id: str, message: dict) -> None:
        """Send message to specific connection."""
        if connection_id in self.active_connections:
            await self.active_connections[connection_id].send_json(message)

    async def broadcast_to_job(self, job_id: str, message: dict) -> None:
        """Broadcast message to all connections subscribed to a job."""
        if job_id in self.job_subscriptions:
            for conn_id in self.job_subscriptions[job_id]:
                if conn_id in self.active_connections:
                    try:
                        await self.active_connections[conn_id].send_json(message)
                    except Exception as e:
                        logger.warning(f"Failed to send to {conn_id}: {e}")

    async def broadcast_all(self, message: dict) -> None:
        """Broadcast to all connected clients."""
        for conn_id, websocket in self.active_connections.items():
            try:
                await websocket.send_json(message)
            except Exception as e:
                logger.warning(f"Failed to broadcast to {conn_id}: {e}")


# Global connection manager
manager = ConnectionManager()

# Shared resources
_watchdog: SelfHealingWatchdog | None = None
_engine: HolographicReconstructionEngine | None = None


def get_watchdog() -> SelfHealingWatchdog:
    """Get or create watchdog instance."""
    global _watchdog
    if _watchdog is None:
        _watchdog = SelfHealingWatchdog()
    return _watchdog


def get_engine() -> HolographicReconstructionEngine:
    """Get or create engine instance."""
    global _engine
    if _engine is None:
        _engine = HolographicReconstructionEngine()
    return _engine


@router.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    """
    Main WebSocket endpoint for real-time communication.

    Message types:
    - subscribe: Subscribe to job updates
    - process: Start image processing with live preview
    - health: Get current system health
    - dev_params: Get/set dev parameters (dev mode)
    """
    connection_id = await manager.connect(websocket)

    try:
        # Send welcome message
        await manager.send_personal(
            connection_id,
            {
                "type": "connected",
                "connection_id": connection_id,
                "message": "Connected to pic2pic-nextgen",
            },
        )

        while True:
            # Receive message
            data = await websocket.receive_text()
            message = json.loads(data)
            msg_type = message.get("type", "")

            if msg_type == "subscribe":
                # Subscribe to job updates
                job_id = message.get("job_id")
                if job_id:
                    manager.subscribe_to_job(connection_id, job_id)
                    await manager.send_personal(
                        connection_id,
                        {"type": "subscribed", "job_id": job_id},
                    )

            elif msg_type == "process":
                # Start processing with live preview
                await handle_process_request(connection_id, message)

            elif msg_type == "health":
                # Get system health
                await handle_health_request(connection_id)

            elif msg_type == "dev_params":
                # Dev mode: get/set parameters
                await handle_dev_params(connection_id, message)

            elif msg_type == "ping":
                # Keepalive
                await manager.send_personal(connection_id, {"type": "pong"})

            else:
                await manager.send_personal(
                    connection_id,
                    {"type": "error", "message": f"Unknown message type: {msg_type}"},
                )

    except WebSocketDisconnect:
        manager.disconnect(connection_id)
    except json.JSONDecodeError:
        await manager.send_personal(
            connection_id,
            {"type": "error", "message": "Invalid JSON"},
        )
        manager.disconnect(connection_id)
    except Exception as e:
        logger.error(f"WebSocket error: {e}")
        manager.disconnect(connection_id)


async def handle_process_request(connection_id: str, message: dict) -> None:
    """Handle image processing request with live preview streaming."""
    import base64

    image_data = message.get("image_data")  # Base64 encoded
    mode_str = message.get("mode", "enhance")

    if not image_data:
        await manager.send_personal(
            connection_id,
            {"type": "error", "message": "No image data provided"},
        )
        return

    try:
        mode = ReconstructionMode(mode_str)
    except ValueError:
        mode = ReconstructionMode.ENHANCE

    # Decode image
    try:
        image_bytes = base64.b64decode(image_data)
    except Exception:
        await manager.send_personal(
            connection_id,
            {"type": "error", "message": "Invalid base64 image data"},
        )
        return

    job_id = str(uuid4())

    # Notify processing started
    await manager.send_personal(
        connection_id,
        {
            "type": "processing_started",
            "job_id": job_id,
            "mode": mode.value,
        },
    )

    # Subscribe to this job
    manager.subscribe_to_job(connection_id, job_id)

    # Get engine and set up progress callback
    engine = get_engine()

    async def progress_callback(progress: ReconstructionProgress):
        """Stream progress updates to WebSocket."""
        shard_data = [
            {
                "shard_id": s.shard_id,
                "scale": s.scale,
                "activation": s.activation,
                "coordinates": s.coordinates,
            }
            for s in progress.shard_activations
        ]

        await manager.broadcast_to_job(
            job_id,
            {
                "type": "progress",
                "job_id": job_id,
                "step": progress.step,
                "total_steps": progress.total_steps,
                "progress_percent": progress.progress_percent,
                "current_scale": progress.current_scale,
                "active_shards": progress.active_shards,
                "shard_activations": shard_data,
            },
        )

    # Wrap sync callback for async
    def sync_progress_callback(progress: ReconstructionProgress):
        asyncio.create_task(progress_callback(progress))

    engine.set_progress_callback(sync_progress_callback)

    # Process image
    try:
        result = await engine.reconstruct(image_bytes, mode)

        # Send completion
        await manager.broadcast_to_job(
            job_id,
            {
                "type": "completed",
                "job_id": job_id,
                "result": base64.b64encode(result).decode(),
            },
        )
    except Exception as e:
        logger.error(f"Processing error: {e}")
        watchdog = get_watchdog()
        watchdog.handle_error(e)

        await manager.broadcast_to_job(
            job_id,
            {
                "type": "error",
                "job_id": job_id,
                "message": str(e),
            },
        )


async def handle_health_request(connection_id: str) -> None:
    """Handle health status request."""
    watchdog = get_watchdog()
    health = await watchdog.check_health()

    await manager.send_personal(
        connection_id,
        {
            "type": "health",
            "status": health.status.value,
            "device_mode": health.device_mode.value,
            "memory_usage_mb": health.memory_usage_mb,
            "cpu_temp_c": health.cpu_temp_c,
            "last_checkpoint": health.last_checkpoint,
            "consecutive_failures": health.consecutive_failures,
            "message": health.message,
        },
    )


async def handle_dev_params(connection_id: str, message: dict) -> None:
    """Handle dev parameter get/set requests."""
    action = message.get("action", "get")
    engine = get_engine()

    if action == "get":
        params = engine.parameters
        await manager.send_personal(
            connection_id,
            {
                "type": "dev_params",
                "action": "get",
                "params": {
                    "tau_scale_1": params.tau_scale_1,
                    "tau_scale_2": params.tau_scale_2,
                    "tau_scale_3": params.tau_scale_3,
                    "tau_scale_4": params.tau_scale_4,
                    "cleanup_k_top": params.cleanup_k_top,
                    "binding_temperature": params.binding_temperature,
                    "binding_strength": params.binding_strength,
                    "cosine_threshold": params.cosine_threshold,
                },
            },
        )
    elif action == "set":
        params_update = message.get("params", {})
        for key, value in params_update.items():
            if hasattr(engine.parameters, key):
                setattr(engine.parameters, key, value)

        await manager.send_personal(
            connection_id,
            {
                "type": "dev_params",
                "action": "set",
                "status": "updated",
            },
        )

File: web-client/src/lib/utils/websocket.ts
/**
 * WebSocket Client
 * ================
 * Manages WebSocket connection for real-time communication with backend.
 */
import {
	wsConnected,
	updateJob,
	shardActivations,
	systemHealth,
	devParameters,
	type ProcessingJob,
	type ShardActivation,
	type SystemHealth,
	type DevParameters
} from '$stores';

let ws: WebSocket | null = null;
let reconnectAttempts = 0;
const maxReconnectAttempts = 10;
const reconnectDelay = 3000;

/**
 * Connect to WebSocket server
 */
export function connect(url?: string): void {
	if (ws?.readyState === WebSocket.OPEN) return;

	const wsUrl = url || getWebSocketUrl();

	try {
		ws = new WebSocket(wsUrl);

		ws.onopen = () => {
			console.log('WebSocket connected');
			wsConnected.set(true);
			reconnectAttempts = 0;
		};

		ws.onclose = () => {
			console.log('WebSocket disconnected');
			wsConnected.set(false);
			ws = null;

			// Auto-reconnect with exponential backoff
			if (reconnectAttempts < maxReconnectAttempts) {
				reconnectAttempts++;
				const delay = reconnectDelay * Math.pow(1.5, reconnectAttempts - 1);
				console.log(`Reconnecting in ${delay}ms...`);
				setTimeout(() => connect(wsUrl), delay);
			}
		};

		ws.onerror = (error) => {
			console.error('WebSocket error:', error);
		};

		ws.onmessage = (event) => {
			handleMessage(JSON.parse(event.data));
		};
	} catch (error) {
		console.error('WebSocket connection failed:', error);
		wsConnected.set(false);
	}
}

/**
 * Disconnect from WebSocket
 */
export function disconnect(): void {
	if (ws) {
		ws.close();
		ws = null;
	}
}

/**
 * Send message through WebSocket
 */
export function send(message: object): void {
	if (ws?.readyState === WebSocket.OPEN) {
		ws.send(JSON.stringify(message));
	} else {
		console.warn('WebSocket not connected');
	}
}

/**
 * Process an image with live preview
 */
export function processImage(imageData: string, mode: string): void {
	send({
		type: 'process',
		image_data: imageData,
		mode
	});
}

/**
 * Subscribe to job updates
 */
export function subscribeToJob(jobId: string): void {
	send({
		type: 'subscribe',
		job_id: jobId
	});
}

/**
 * Request health status
 */
export function requestHealth(): void {
	send({ type: 'health' });
}

/**
 * Get dev parameters
 */
export function getDevParams(): void {
	send({
		type: 'dev_params',
		action: 'get'
	});
}

/**
 * Set dev parameters
 */
export function setDevParams(params: Partial<DevParameters>): void {
	send({
		type: 'dev_params',
		action: 'set',
		params
	});
}

/**
 * Handle incoming WebSocket message
 */
function handleMessage(message: { type: string; [key: string]: unknown }): void {
	switch (message.type) {
		case 'connected':
			console.log('Connected with ID:', message.connection_id);
			break;

		case 'processing_started':
			updateJob({
				id: message.job_id as string,
				status: 'processing',
				progress: 0,
				mode: message.mode as string
			});
			break;

		case 'progress':
			updateJob({
				id: message.job_id as string,
				status: 'processing',
				progress: message.progress_percent as number,
				mode: ''
			});
			shardActivations.set(message.shard_activations as ShardActivation[]);
			break;

		case 'completed':
			updateJob({
				id: message.job_id as string,
				status: 'completed',
				progress: 100,
				mode: '',
				resultUrl: `data:image/png;base64,${message.result}`
			});
			break;

		case 'error':
			if (message.job_id) {
				updateJob({
					id: message.job_id as string,
					status: 'error',
					progress: 0,
					mode: '',
					error: message.message as string
				});
			}
			console.error('Error:', message.message);
			break;

		case 'health':
			systemHealth.set({
				status: message.status as SystemHealth['status'],
				device_mode: message.device_mode as SystemHealth['device_mode'],
				memory_usage_mb: message.memory_usage_mb as number,
				cpu_temp_c: message.cpu_temp_c as number | null,
				message: message.message as string
			});
			break;

		case 'dev_params':
			if (message.action === 'get') {
				devParameters.set(message.params as DevParameters);
			}
			break;

		case 'pong':
			// Keepalive response
			break;

		default:
			console.log('Unknown message type:', message.type);
	}
}

/**
 * Get WebSocket URL based on current location
 */
function getWebSocketUrl(): string {
	const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
	const host = window.location.host;

	// In development, connect to backend directly
	if (import.meta.env.DEV) {
		return 'ws://localhost:8000/ws';
	}

	return `${protocol}//${host}/ws`;
}

// Ping to keep connection alive
setInterval(() => {
	if (ws?.readyState === WebSocket.OPEN) {
		send({ type: 'ping' });
	}
}, 30000);
